caffe2/utils/math_utils.cc

```diff
@@ -0,0 +1,141 @@
+#include "caffe2/utils/math_utils.h"
+
+#include <algorithm>
+#include <vector>
+
+#include "caffe2/core/logging.h"
+
+namespace caffe2 {
+namespace math {
+namespace utils {
+
+void IncreaseIndexInDims(const int n, const int* dims, int* index) {
+  for (int i = n - 1; i >= 0; --i) {
+    ++index[i];
+    if (index[i] >= dims[i]) {
+      index[i] -= dims[i];
+    } else {
+      break;
+    }
+  }
+}
+
+int GetIndexFromDims(const int n, const int* dims, const int* index) {
+  int sum = 0;
+  for (int i = 0; i < n; ++i) {
+    if (dims[i] > 1) {
+      sum = sum * dims[i] + index[i];
+    }
+  }
+  return sum;
+}
+
+void ComputeBroadcastBinaryOpDims(
+    const int A_ndim,
+    const int* A_dims,
+    const int B_ndim,
+    const int* B_dims,
+    int* A_broadcast_dims,
+    int* B_broadcast_dims,
+    int* C_broadcast_dims) {
+  const int ndim = std::max(A_ndim, B_ndim);
+  std::fill(A_broadcast_dims, A_broadcast_dims + ndim - A_ndim, 1);
+  std::fill(B_broadcast_dims, B_broadcast_dims + ndim - B_ndim, 1);
+  std::copy(A_dims, A_dims + A_ndim, A_broadcast_dims + ndim - A_ndim);
+  std::copy(B_dims, B_dims + B_ndim, B_broadcast_dims + ndim - B_ndim);
+  for (int i = 0; i < ndim; ++i) {
+    CAFFE_ENFORCE(
+        A_broadcast_dims[i] == B_broadcast_dims[i] ||
+        A_broadcast_dims[i] <= 1 || B_broadcast_dims[i] <= 1);
+    if (A_broadcast_dims[i] == 0 || B_broadcast_dims[i] == 0) {
+      C_broadcast_dims[i] = 0;
```

Comment:
Just for my own education, what does dim=0 mean? How do we handle this situation?

Reply:
If some dim == 0, that means the tensor is empty while they have the dimensions. I just make this to be the same behavior as numpy and support some of the sparse features usecase.